cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_single_col.hpp

```diff
@@ -0,0 +1,154 @@
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+// Performs k-selection for k == 1
+template <typename Float>
+class kselect_by_rows_single_col : public kselect_by_rows_base<Float> {
+public:
+    kselect_by_rows_single_col() {}
+    sycl::event operator()(sycl::queue& queue,
+                           const ndview<Float, 2>& data,
+                           std::int64_t k,
+                           ndview<Float, 2>& selection,
+                           ndview<std::int32_t, 2>& indices,
+                           const event_vector& deps) override {
+        return select<true, true>(queue, data, selection, indices, deps);
+    }
+    sycl::event operator()(sycl::queue& queue,
+                           const ndview<Float, 2>& data,
+                           std::int64_t k,
+                           ndview<Float, 2>& selection,
+                           const event_vector& deps) override {
+        ndarray<std::int32_t, 2> dummy;
+        return select<true, false>(queue, data, selection, dummy, deps);
+    }
+    sycl::event operator()(sycl::queue& queue,
+                           const ndview<Float, 2>& data,
+                           std::int64_t k,
+                           ndview<std::int32_t, 2>& indices,
+                           const event_vector& deps) override {
+        ndarray<Float, 2> dummy;
+        return select<false, true>(queue, data, dummy, indices, deps);
+    }
+
+private:
+    template <bool selection_out, bool indices_out>
+    sycl::event select(sycl::queue& queue,
+                       const ndview<Float, 2>& data,
+                       ndview<Float, 2>& selection,
+                       ndview<std::int32_t, 2>& indices,
+                       const event_vector& deps = {}) {
+        if (indices_out) {
+            ONEDAL_ASSERT(indices.get_shape()[0] == data.get_shape()[0]);
+            ONEDAL_ASSERT(indices.get_shape()[1] == 1);
+        }
+        if (selection_out) {
+            ONEDAL_ASSERT(selection.get_shape()[0] == data.get_shape()[0]);
+            ONEDAL_ASSERT(selection.get_shape()[1] == 1);
+        }
+        const auto sg_sizes = queue.get_device().get_info<sycl::info::device::sub_group_sizes>();
+        ONEDAL_ASSERT(!sg_sizes.empty());
+
+        auto result = std::max_element(sg_sizes.begin(), sg_sizes.end());
+        ONEDAL_ASSERT(result != sg_sizes.end());
+
+        const std::int64_t sg_max_size = static_cast<std::int64_t>(*result);
+
+        const std::int64_t col_count = data.get_dimension(1);
+        const std::int64_t row_count = data.get_dimension(0);
+        const std::int64_t stride = data.get_shape()[1];
+
+        const std::int64_t row_adjusted_sg_num =
+            col_count / sg_max_size + std::int64_t(col_count % sg_max_size > 0);
+        const std::int64_t expected_sg_num =
+            std::min(kselect_by_rows_single_col::preffered_wg_size / sg_max_size,
+                     row_adjusted_sg_num);
+        ONEDAL_ASSERT(expected_sg_num > 0);
```

Comment:
I see that many code snippets looks similar or even copy-pasted to what we have in cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_dpc.cpp
Can we extract common things from these files?